### DISCUSSION TITLE
Rate limiting: Query depth and Requests per minute

### DISCUSSION THREAD

**Discussion by cvb941** [nhost/nhost#1688] [Q&A]
Hi, is it possible to set up, or does Nhost already apply some rate limiting for GraphQL query depth and requests per minute?

Hasura [has this feature](https://hasura.io/learn/graphql/hasura-advanced/security/4-rate-limit/), but it is only included in their Pro tier.

We consider rate limiting a critical security feature that stops us from using Nhost.

**Comment by ammarfaris** [comment]
+1

**Comment by dbarrosop** [comment]
Not available for free/pro. You can integrate [stellate](https://stellate.co) and get this and other features or we could do it for you as part of an enterprise plan.

**Reply by ammarfaris** [reply]
Thanks for your swift response @dbarrosop. Stellate looks great as a temporary alternative albeit the rate-limiting feature is still in "private alpha". 

Just wondering though, for Stellate, if a user somehow managed to get a hold of my GraphQL endpoint origin (the original graphql link), will it be able to bypass Stellate's link / services ? 

Just thinking, say for its rate-limiting feature, if someone bypass the "Stellate Edge Cache subdomain" (and use original graphql link), it will become less effective right? unless the original link is long-and-random and never exposed before, will it be fine? not sure if i'm thinking this correctly with Stellate as an intermediary that sits in between the user's request and Hasura / Nhost. 

Thanks in advance!

**Reply by dbarrosop** [reply]
Yes, the CDN can be bypassed if you know the original URL but rate-limiting on the service side has other considerations. Keep in mind these features are more often aimed at improving performance and protecting yourself from application bugs (imagine a for loop trying to fetch data without any backoff timer).

If you are concerned about attackers, you need to protect yourself with a CDN or similar service. If you try to rate-limit on the service-side (or in the cloud itself), a volumetric DDoS attack can easily overwhelm the service rendering it unable to serve requests (i.e. all of its computing capacity is being used to simply check if requests are allowed or not) or, more likely, simply saturate the network.

So it all depends on what your risk profile is and what your concerns are and if you are in need of customized solutions we can take your requirements in consideration as part of an enterprise plan.

**Reply by ammarfaris** [reply]
Thanks @dbarrosop for a comprehensive explanation. Would like to ask though, for the Nhost Hosted Platform, is there any sort of protection for these kind of attacks? One concern is that it may result in high network bandwidth coming from massive malicious API calls which result in surprising charges / fees? Or is there some sort protection / "spend cap" ? Thanks

**Reply by dbarrosop** [reply]
We don't charge for bandwidth although we have plans to change that very soon. In any case, DDoS attacks are expensive for attackers so unless your application has a high-risk profile I wouldn't worry about it and if your application does have a high-risk profile I'd suggest contacting us for enterprise/customized plan where we could look at your security/risk profile.

**Reply by ammarfaris** [reply]
okay good insights. thanks again for taking the time to respond to my queries, @dbarrosop .
